### PR TITLE
Adds CloudinaryImage(s) fixes, video support (beta), select (beta), opinionated enhancements. #832, #3504, #2699

### PR DESCRIPTION
**Pull request by danielmahon**
## Description of changes
CloudinaryImage(s) fixes, video support (beta), select (beta), opinionated enhancements.
~~Fixes folder support for CloudinaryImages. Previously only worked with single image.~~
*This PR has outgrown its original purpose... they grow up so fast 😉 *

- [x] Adds `autoCleanup` support to CloudinaryImage(s)
- [x] Adds `public_id` `format` and `resource_type` to CloudinaryImages lightbox
- [x] Adds the `etag` property, which is the MD5 generated by Cloudinary on upload. Useful for local file comparisons.
- [x] Adds `select` functionality, allowing for choosing remote Cloudinary Images
  - [x] Update select's UI with preview images  
  - [ ] Support CloudinaryImages (plural)
  - [x] Currently shows "sibling" uploaded media with prefix format:
         `[cloudinary prefix]/[list path]/[field path]/`
- [x] Adds video upload/preview support (beta) to CloudinaryImage
  - [x] Allow and support video uploads
  - [x] Generate video thumbnails in admin
  - [x] Adds video icon to thumbnails in admin
  - [ ] Support videos in CloudinaryImages (plural) as well
  - [ ] Allow video playback in admin
  - [ ] Handle video use cases for templates (I only needed uploads for time being)

## Questions
Am I wasting my time? 🌳🔨  
Should we refactor CloudinaryImage(s) to work as one field type instead of two?
Rename field to something like `CloudinaryUpload || CloudinaryMedia || Cloudinary`?

## Screenshots
Supports a mixing of images and video:
hydrochem-panel-server

Supports upload or selecting from Cloudinary:
hydrochem

Adds image previews, file-types, and paths to dropdown:
hydrochem

Supports metadata in lightbox caption:
cursor_and_hydrochem-panel-server

Adds "open original" button to lightbox:
hydrochem

## Related issues (if any)
#832
#3504 
#2699

## Testing

- [x] Please confirm `npm run test-all` ran successfully.

### CODE REVIEW

**Comment by danielmahon** [comment]
I don't think I'm missing anything here but I haven't looked at the 4.x codebase for a bit so while it should be a simple fix, maybe just look it over.

**Comment by danielmahon** [comment]
I just realized that Cloudinary might be getting replaced by the Storage API... is that correct? Are 4.x changes needed to Cloudinary types?

**Comment by danielmahon** [comment]
Just added support (without trying to change too much) for autoCleanup removal of single and multiple images.

**Comment by danielmahon** [comment]
On another note, I would like to integrate video upload support to this as well, as Cloudinary now supports it.

**Comment by danielmahon** [comment]
@JedWatson I've started to add video upload support as well. There aren't THAT many changes to make this work but it does complicate this pull-request a tad. Let me know if you'd like me to handle this a different way. My internal thought was to add video support (behind the scenes) and then maybe later we could do a refactor of CloudinaryImages to something like CloudinaryUpload and try to get it to handle "all the things". I've got a project I'm trying to use this with ASAP hence me throwing these things together, but I'd be down to refactor later. It'd be nice to remove the need for 2 field types.

File: fields/types/cloudinaryimage/CloudinaryImageType.js [hunk 1]
```diff
@@ -370,8 +370,8 @@ cloudinaryimage.prototype.updateItem = function (item, data, files, callback) {
 	var value = this.getValueFromData(data);
 	var uploadedFile;
 
-	// Providing the string "remove" removes the file and resets the field
-	if (value === 'remove') {
+	// Providing the string "upload:remove" removes the file and resets the field
+	if (field.options.autoCleanup && value === 'upload:remove') {
```
Comment (JedWatson) [inline fields/types/cloudinaryimage/CloudinaryImageType.js]: What's the reason for this change?

File: fields/types/cloudinaryimages/CloudinaryImagesType.js [hunk 2]
```diff
@@ -307,8 +296,20 @@ cloudinaryimages.prototype.updateItem = function (item, data, files, callback) {
 
 	async.map(values, function (value, next) {
 		if (typeof value === 'object' && 'public_id' in value) {
+			// Setting "autoCleanup" the property "remove:true" removes the file and resets the field
```
Comment (JedWatson) [inline fields/types/cloudinaryimages/CloudinaryImagesType.js]: Do you mean 'Setting "autoCleanup" **or** the property' here?

File: fields/types/cloudinaryimage/CloudinaryImageField.js [hunk 3]
```diff
@@ -292,7 +295,7 @@ module.exports = Field.create({
 		if (this.state.userSelectedFile || this.state.removeExisting) {
 			const value = this.state.userSelectedFile
 				? `upload:${this.state.uploadFieldPath}`
-				: '';
+				: `upload:remove`;
```
Comment (JedWatson) [inline fields/types/cloudinaryimage/CloudinaryImageField.js]: Good catch, but I think checking for `remove` (mirrored in `CloudinaryImageType.js` below) is probably clearer

File: fields/types/cloudinaryimages/CloudinaryImagesField.js [hunk 4]
```diff
@@ -16,8 +16,8 @@ import Thumbnail from './CloudinaryImagesThumbnail';
 import HiddenFileInput from '../../components/HiddenFileInput';
 import FileChangeMessage from '../../components/FileChangeMessage';
 
-const SUPPORTED_TYPES = ['image/*', 'application/pdf', 'application/postscript'];
-const SUPPORTED_REGEX = new RegExp(/^image\/|application\/pdf|application\/postscript/g);
+const SUPPORTED_TYPES = ['video/*', 'image/*', 'application/pdf', 'application/postscript'];
```
Comment (JedWatson) [inline fields/types/cloudinaryimages/CloudinaryImagesField.js]: This should also be applied to `CloudinaryImageField.js`, right?

**Comment by Moenomi** [comment]
Is this already merged? I just stumbled over this Issue since the autoCleanup of the CloudinaryImage is currently not working in the latest version. I'm quite new to Keystone but after looking at the code and testing I found that the value is never 'remove' in CloudinaryImageType.js.

File: fields/types/cloudinaryimages/CloudinaryImagesType.js (hunk 2 above)
Comment (danielmahon) [inline fields/types/cloudinaryimages/CloudinaryImagesType.js]: @JedWatson finally getting back to this! Ill change the the comment to:
```
// If "autoCleanup" is enabled and the "remove" property is true for "value"
// than remove the file and reset the field
```
Does that clear it up?

File: fields/types/cloudinaryimage/CloudinaryImageField.js (hunk 3 above)
Comment (danielmahon) [inline fields/types/cloudinaryimage/CloudinaryImageField.js]: @JedWatson for some reason when I set just `remove` I get a 400 POST error. It is expecting that `upload:` string... Something to do with express or something else? I haven't tracked down the problem yet.

File: fields/types/cloudinaryimages/CloudinaryImagesField.js (hunk 4 above)
Comment (danielmahon) [inline fields/types/cloudinaryimages/CloudinaryImagesField.js]: @JedWatson I hadn't finished working out video support on the single CloudinaryImage yet. Will work on it soon.

File: fields/types/cloudinaryimage/CloudinaryImageType.js (hunk 1 above)
Comment (danielmahon) [inline fields/types/cloudinaryimage/CloudinaryImageType.js]: @JedWatson I am currently getting a 400 POST error if I don't include the `upload:` prefix

**Comment by danielmahon** [comment]
Hey @JedWatson, just checking in to see that these changes are still on track. I'm hoping to have `select` up and running today. Tired of having so many duplicates in Cloudinary ;)

**Comment by danielmahon** [comment]
@JedWatson Added select functionality to CloudinaryImage and UI updates, take a look!

File: fields/types/cloudinaryimage/CloudinaryImageType.js (hunk 1 above)
Comment (danielmahon) [inline fields/types/cloudinaryimage/CloudinaryImageType.js]: @JedWatson Figured this out, it was failing validation, needs to be set to `delete`. See https://github.com/keystonejs/keystone/blob/master/fields/types/cloudinaryimage/CloudinaryImageType.js#L293

File: fields/types/cloudinaryimage/CloudinaryImageField.js (hunk 3 above)
Comment (danielmahon) [inline fields/types/cloudinaryimage/CloudinaryImageField.js]: @JedWatson Figured this out, it was failing validation, needs to be set to delete. See https://github.com/keystonejs/keystone/blob/master/fields/types/cloudinaryimage/CloudinaryImageType.js#L293

**Comment by MillerApps** [comment]
Is this still in review?

**Comment by Noviny** [comment]
Going to add this to our new [PR statuses](https://github.com/keystonejs/keystone/projects/3) project, and assign it to someone to look through.

**Comment by danielmahon** [comment]
@Noviny Awesome Ben! I am using this/my branch in a few production (low traffic) projects so after making the changes I needed this PR got a little out of hand and some it might be too "opinionated" about how it handles things. I think ultimately it would be best to refactor the entire CloudinaryImage(s) type into a single type like "Cloudinary" or "CloudinaryMedia" since I've been using it to handle documents/images/videos. I would be happy to work on that once I know you guys are on board with that direction or if another implementation would be better. Let me know. Thanks!

**Comment by morenoh149** [comment]
@danielmahon its inspiring that you're still improving this pr. I'm sure it'll be merged soon!

**Comment by gusgard** [comment]
@danielmahon Thanks, save my day!

**Comment by danielmahon** [comment]
aging-season-9-gif-by-friends
